### PR TITLE
Support configuration cache

### PR DESCRIPTION
**Pull request by sebek64**
Support configuration cache.

This is the current standard for plugins. It makes the builds faster by supporting heavy parallelism in builds, and slow whenever a non-compliant plugin is part of the build.

I tried carefully to set all inputs / outputs of tasks. It is however quite hard to handle all imaginable cases.

The commits should be small and reviewable. If anything is not clear from the commits, I'm ready to explain.

### CODE REVIEW

**Comment by sebek64** [comment]
I already found one case where it doesn't work correctly, so a fix should be added soon (hopefully next Monday).

**Comment by sebek64** [comment]
The fix has been pushed - bufTool configuration creation before task creation.
With the new inputs/outputs configured, it can be a bit tricky to use this plugin in combination with other plugins like java(-library), in particular its `processResources` task. The user has to explicitly specify dependency of BufLint / BufFormatCheck (and possibly others) on the `processResources` task.
Because of that, if this PR gets merged, I suggest releasing the plugin with new "major" version (0.10.0?), to make the compatibility breaking obvious.

**Comment by andrewparmet** [comment]
Thanks for all the work in this contribution!

Is the `processResources` dependency the only complication for users that this PR introduces? Did that dependency always exist and the plugin just happened to work?

I think requiring users to declare that dependency is a non-starter - we should try as hard as we can not to require users to do any task dependency declarations like that.

**Comment by sebek64** [comment]
Thanks for quick reaction. After a few days of usage of this "modified" plugin version, I think I understand the traps and problems a bit more.
In the current version, the tasks do not specify any inputs/outputs. So whenever the task is requested, it is always run. You can see that on info logging level:
```
./gradlew bufBuild --info
...
Task ':...:bufBuild' is not up-to-date because:
  Task has not declared any outputs despite executing actions.
Running buf from ...: `buf build --output ...build/bufbuild/image.json`
```
In the new version, the inputs/outputs are declared, so the buf is only executed if the inputs change. So for example
```
./gradlew BufBuild --info
...
> Task :...:bufBuild UP-TO-DATE
Caching disabled for task ':...:bufBuild' because:
  Build cache is disabled
  Caching has not been enabled for the task
Skipping task ':...:bufBuild' as it is up-to-date.
```
Now we come to the fact that buf tool itself is not very compatible with gradle way of organizing repositories. The standard gradle way is to have two directories in the module - `src` and `build`. Simply speaking, task inputs are usually in `src`, task outputs are usually in `build`.
Sometimes, tasks can have inputs in `build`. Then the tasks are chained and explicit dependencies between them is needed. A real-world example is `processResources` task, reading some files from `src/.../resources`, and storing them to `build/.../resources`. Then some other tasks like `jar` have `build/.../resources` as the inputs. Therefore, `jar` has to depend on `processResources`.
Sometimes, tasks can have outputs in `src` directory. For example, ktlint (kotlin linter) has a `ktlintFormat` task, which applies linting rules to existing sources. This task has outputs in the `src` directory.
With buf directory layout, the sources are placed directly into the root tree (or at least that's one of the options). Then the inputs are basically taken from there. It is possible to have inputs even in the `build` directory, leading to the need of explicit specification of task dependencies.
In this PR, I explicitly excluded `build` from task inputs. Then the build doesn't fail because of missing task dependencies. But it can possibly incorrectly skip a task run if the inputs are in the build directory (outputs of some other task), but not declared explicitly.
Part of the problem is that this plugin is not configured entirely inside gradle, but partially also via external files (buf.yaml). I'm afraid there is no perfect solution, except for doing even bigger refactoring, and encapsulating all the configuration (at least regarding inputs/outputs) into the gradle itself, and generating `buf.yaml` as part of plugin execution.
I'm open to discussion. We currenly use a forked version for our needs, so we may polish this PR and take time with it.

**Comment by andrewparmet** [comment]
Thanks for the detailed explanation. I guess phrased differently my question is: what is it about this change that makes the plugin now depend on `processResources`?

**Comment by sebek64** [comment]
Let me reply shortly this time. There is nothing that makes this plugin depend on `processResources`. The first comment mentioning this task was before the `build` directory was added to excludes (see the last commit in this PR).

File: src/test/kotlin/build/buf/gradle/AbstractBufIntegrationTest.kt [hunk 1]
```diff
@@ -70,6 +70,7 @@ abstract class AbstractBufIntegrationTest : IntegrationTest {
                 "-PprotobufVersion=3.23.4",
                 "-PkotlinVersion=1.7.20",
                 "-PandroidGradleVersion=7.3.0",
+                "--configuration-cache",
```
Comment (andrewparmet) [inline src/test/kotlin/build/buf/gradle/AbstractBufIntegrationTest.kt:73]: Nice dogfooding.

File: src/test/resources/GenerateTest/buf_generate_fails_with_both_default_and_specified_buf_gen_template_files/build.gradle [hunk 2]
```diff
@@ -2,6 +2,10 @@ plugins {
   id 'build.buf'
 }
 
+repositories {
```
Comment (andrewparmet) [inline src/test/resources/GenerateTest/buf_generate_fails_with_both_default_and_specified_buf_gen_template_files/build.gradle:5]: Why is this necessary now (and seemingly only for some tests)? If it is necessary, then I think it should be added to the example code (or at least a blurb addressing the error that occurs if it is absent) in the readme as well.

File: src/main/kotlin/build/buf/gradle/AbstractBufTask.kt [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.Input
+import java.io.File
+import java.nio.file.Path
+
+abstract class AbstractBufTask : DefaultTask() {
+    // this is just an @Input, not @InputDirectory
+    // this property is used only for relativization of paths, the task doesn't depend on the content
+    @get:Input
+    abstract val projectDir: Property<File>
```
Comment (andrewparmet) [inline src/main/kotlin/build/buf/gradle/AbstractBufTask.kt:27]: Is there any way to restrict the visibility of all of these properties on the various task classes? I'd rather them not become part of the plugin's public API.

Also, while it's fresh, I think we should get proper KDoc on each property (even if we can make them internal). I'll never remember what they are a year from now...

File: src/main/kotlin/build/buf/gradle/BuildConfiguration.kt [hunk 4]
```diff
@@ -28,16 +27,22 @@ private const val BUF_BUILD_PUBLICATION_FILE_BASE_NAME = "image"
...
 
         if (hasProtobufGradlePlugin()) {
             dependsOn(COPY_BUF_CONFIG_TASK_NAME)
-        } else {
-            // Called already during workspace configuration if the protobuf-gradle-plugin has been applied
-            createsOutput()
         }
+
+        inputFiles.setFrom(
+            fileTree(workingDir.get()) {
```
Comment (andrewparmet) [inline src/main/kotlin/build/buf/gradle/BuildConfiguration.kt:39]: This feels a bit spaghetti-ish to me with various `get()` calls distanced from their `set()` calls, but perhaps I just have to adjust to a different syntax style.

File: src/main/kotlin/build/buf/gradle/FormatConfiguration.kt [hunk 5]
```diff
@@ -27,18 +27,41 @@ internal fun Project.configureFormat() {
...
+    registerBufExecTask<FormatApplyTask>(BUF_FORMAT_APPLY_TASK_NAME) {
         group = VERIFICATION_GROUP
         description = "Formats a Protobuf schema according to Buf's formatting rules."
+
+        inputFiles.setFrom(
+            fileTree(workingDir.get()) {
+                include("**/*.proto")
+                // not to interfere with random plugins producing output to build dir
+                exclude("build")
+            },
```
Comment (andrewparmet) [inline src/main/kotlin/build/buf/gradle/FormatConfiguration.kt:57]: This seems to be a pretty common code block. Do you anticipate it changing across invocations, or can it be extracted to a helper function somewhere?

File: src/main/kotlin/build/buf/gradle/ProtobufGradlePluginSupport.kt [hunk 6]
```diff
@@ -96,21 +125,14 @@ private fun Task.workspaceCommonConfig() {
...
-        .map { project.makeMangledRelativizedPathStr(it) }
-        .joinToString("\n") { "|  - $it" }
-
 // Returns all directories that have may have proto files relevant to processing the project's proto files. This
 // includes any proto files that are simply references (includes) as well as those that will be processed (code
 // generation or validation).
 private fun Task.allProtoDirs() =
     project.allProtoSourceSetDirs()
         .plus(project.file(Paths.get(BUILD_EXTRACTED_INCLUDE_PROTOS_MAIN)))
-        .filter { anyProtos(it) }
```
Comment (andrewparmet) [inline src/main/kotlin/build/buf/gradle/ProtobufGradlePluginSupport.kt:113]: Why are these filters removed? If I remember correctly, this would cause a buf invocation to fail if any source directories didn't contain proto files. I hope there's a test that would fail if the only change to `main` was the removal of these filters, but I can't remember.

I see the `anyProtos` function is called elsewhere. Am I just not seeing the whole picture where the filter is applied later?

**Review by andrewparmet — COMMENTED** [review]
Sorry for the delay here. @drice-buf for visibility.

Ok, I think I see what you're saying - where previously there was some awkwardness, you've now arrived at a place where there is no complexity imparted upon the user of the plugin to support the configuration cache. That's great news, and I think it is a Pareto improvement so long as there are few edge cases you haven't considered. To be honest, the Gradle configuration cache is an area I know little about (perhaps this is obvious), so bear with me as I try to get up to speed on this PR.

File: src/main/kotlin/build/buf/gradle/BuildConfiguration.kt (hunk 4 above)
Comment (sebek64) [reply src/main/kotlin/build/buf/gradle/BuildConfiguration.kt:39]: This is gradle api, I didn't invent it. I don't know it so well, so maybe there is a better way to use it.

File: src/main/kotlin/build/buf/gradle/FormatConfiguration.kt (hunk 5 above)
Comment (sebek64) [reply src/main/kotlin/build/buf/gradle/FormatConfiguration.kt]: Extracted to a function in the last commit.

File: src/test/resources/GenerateTest/buf_generate_fails_with_both_default_and_specified_buf_gen_template_files/build.gradle (hunk 2 above)
Comment (sebek64) [reply src/test/resources/GenerateTest/buf_generate_fails_with_both_default_and_specified_buf_gen_template_files/build.gradle:5]: If I remember correctly, it's because of reordering steps in which they're executed. These tests test some kind of failures. Originally, these failures happened before buf executable was resolved. After the change, buf executable is resolved before this failure occurs, so the test fails on something different. Providing the repo makes the test reach the desired failure.

File: src/main/kotlin/build/buf/gradle/ProtobufGradlePluginSupport.kt (hunk 6 above)
Comment (sebek64) [reply src/main/kotlin/build/buf/gradle/ProtobufGradlePluginSupport.kt:113]: These calls are just moved around the code. See this commit (or all commits in the sequence). There should not be any behavior change. If so, then I made a mistake.

File: src/main/kotlin/build/buf/gradle/AbstractBufTask.kt (hunk 3 above)
Comment (sebek64) [reply src/main/kotlin/build/buf/gradle/AbstractBufTask.kt:29]: Added a few kdocs, although their usefulness is a bit low.
Not sure about the way to hide the inputs/outputs - kotlin internal doesn't work, Internal annotation in gradle has different meaning and semantics.

Comment (andrewparmet) [reply src/main/kotlin/build/buf/gradle/AbstractBufTask.kt:29]: There's an `@Internal` Gradle annotation - I wonder what that does.

E.g. https://github.com/google/protobuf-gradle-plugin/blob/master/src/main/groovy/com/google/protobuf/gradle/ProtobufExtract.groovy#L71

File: src/main/kotlin/build/buf/gradle/AbstractBufExecTask.kt [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+import java.io.File
+import java.nio.charset.StandardCharsets
+
+/**
+ * A task executing buf executable as part of its operation.
+ */
+abstract class AbstractBufExecTask : AbstractBufTask() {
+    /** The buf executable. */
+    @get:InputFiles
+    abstract val bufExecutable: ConfigurableFileCollection
```
Comment (andrewparmet) [inline src/main/kotlin/build/buf/gradle/AbstractBufExecTask.kt:30]: Just noticed this - what makes this `InputFiles` and a `ConfigurableFileCollection` instead of `InputFile` and a `Property<File>`?

File: src/main/kotlin/build/buf/gradle/AbstractBufTask.kt (hunk 3 above)
Comment (andrewparmet) [reply src/main/kotlin/build/buf/gradle/AbstractBufTask.kt:29]: > although their usefulness is a bit low.

I happen to think the nontrivial ones (and even some of the simpler ones) will be helpful and establish good practice. Thanks for adding them!

File: src/main/kotlin/build/buf/gradle/ProtobufGradlePluginSupport.kt (hunk 6 above)
Comment (andrewparmet) [reply src/main/kotlin/build/buf/gradle/ProtobufGradlePluginSupport.kt:113]: I've verified that my hope is reality - there are tests that fail if either or both of these filter calls are removed.

File: src/main/kotlin/build/buf/gradle/AbstractBufExecTask.kt [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+    /** Directories possibly containing input .proto files. */
+    @get:InputFiles
+    abstract val candidateProtoDirs: ConfigurableFileCollection
+
+    /** Whether the project has buf workspace or not. */
+    @get:Input
+    abstract val hasWorkspace: Property<Boolean>
+
+    internal fun execBuf(
```
Comment (andrewparmet) [inline src/main/kotlin/build/buf/gradle/AbstractBufExecTask.kt:52]: I think I'd prefer keeping these utility functions in their separate files as extension functions. They don't really fit the "traditional" abstract class pattern of, for example, calling into overrideable methods, and I think it's nice to keep the abstract classes as simple containers of properties.

File: src/main/kotlin/build/buf/gradle/BufSupport.kt [hunk 9]
```diff
@@ -15,11 +15,14 @@
 package build.buf.gradle
 
 import org.gradle.api.Project
-import org.gradle.api.Task
-import java.nio.charset.StandardCharsets
+import org.gradle.kotlin.dsl.dependencies
 
 const val BUF_BINARY_CONFIGURATION_NAME = "bufTool"
 
+internal fun Project.precreateBufDependency() {
```
Comment (andrewparmet) [inline src/main/kotlin/build/buf/gradle/BufSupport.kt:22]: Would this be more accurately named `createBufBinaryDependencyConfiguration` or `createBufBinaryConfiguration`?

File: src/main/kotlin/build/buf/gradle/AbstractBufTask.kt (hunk 3 above)
Comment (sebek64) [reply src/main/kotlin/build/buf/gradle/AbstractBufTask.kt:29]: For gradle `@Internal`, see the docs: https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/Internal.html
In particular, "This will cause the task not to be considered out-of-date when the property has changed.", which is not what we want.

Comment (sebek64) [reply src/main/kotlin/build/buf/gradle/AbstractBufTask.kt:29]: The issue with kotlin `internal` is that I cannot put that on parent classes (`AbstractBufExecTask`, `AbstractBufTask`) - I don't know exactly why. Putting it just on the concrete classes seems very inconsistent.

File: src/main/kotlin/build/buf/gradle/AbstractBufExecTask.kt (hunk 7 above)
Comment (sebek64) [reply src/main/kotlin/build/buf/gradle/AbstractBufExecTask.kt:30]: I used `ConfigurableFileCollection` to make it easily set from "gradle configuration" (see the setting of this property).

File: src/main/kotlin/build/buf/gradle/BufSupport.kt (hunk 9 above)
Comment (sebek64) [reply src/main/kotlin/build/buf/gradle/BufSupport.kt]: Changed.

File: src/main/kotlin/build/buf/gradle/AbstractBufExecTask.kt (hunk 8 above)
Comment (sebek64) [reply src/main/kotlin/build/buf/gradle/AbstractBufExecTask.kt]: I've reworked all commits to comply to this rule (and to make as minimal changes in the structure as possible).

Comment (andrewparmet) [reply src/main/kotlin/build/buf/gradle/AbstractBufExecTask.kt]: Looks great - thanks!

File: src/main/kotlin/build/buf/gradle/AbstractBufTask.kt (hunk 3 above)
Comment (sebek64) [reply src/main/kotlin/build/buf/gradle/AbstractBufTask.kt:28]: When I change `AbstractBufTask.projectDir` to internal, the test fail on
```
FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':publishBufImagePublicationPublicationToMavenRepository'.
> Could not create task ':bufBuild'.
   > 'org.gradle.api.provider.Property build.buf.gradle.AbstractBufTask.getProjectDir()'
```

File: src/main/kotlin/build/buf/gradle/AbstractBufExecTask.kt (hunk 7 above)
Comment (andrewparmet) [reply src/main/kotlin/build/buf/gradle/AbstractBufExecTask.kt:29]: It saves the call to `singleFile` - is that right? 

`bufExecutable.set(configurations.getByName(BUF_BINARY_CONFIGURATION_NAME).singleFile`

But you end up having to call that later. Perhaps having a `Property<File>` is more clear.

Comment (sebek64) [reply src/main/kotlin/build/buf/gradle/AbstractBufExecTask.kt:29]: If I remember correctly, I cannot just call `singleFile` at that point, because the dependency is not yet resolved. So later it is resolved, the better.

Comment (andrewparmet) [reply src/main/kotlin/build/buf/gradle/AbstractBufExecTask.kt:29]: Ah, makes sense.

**Comment by andrewparmet** [comment]
Looks good to me, but a single test is failing on Windows.

**Comment by sebek64** [comment]
Is there a way to see which test / how it failed?

**Comment by andrewparmet** [comment]
CI used to publish the results to the actions summary and it looks like that broke at some point during maintenance. 

I would try using Gradle test logging events?

**Comment by sebek64** [comment]
Ok, the failing test is `GenerateTest > generate java with --include-imports()` - failing on `bufLint`. The crash is quite strange, almost like some out-of-memory in the test worker. Not sure what to do with that.

**Comment by andrewparmet** [comment]
@sebek64 I've done my best to merge in the main branch here but there's some semantic conflicts with Buf's v2 configurations. I'm pausing here but feel free to step in or I'll come back later.

**Comment by sebek64** [comment]
I have already a rebase in progress locally, but I wasn't able to finish it yet. I'll come back to it next week.

**Comment by sebek64** [comment]
Rebased & adapted to changes. Also merged the commits added during previous reviews into correct commits.

**Comment by andrewparmet** [comment]
Looks like the failure in the Windows test is related to rate limiting:

```
2024-08-26T16:39:45.4073581Z GenerateWithWorkspaceTest > buf generate with buf gen template file override() FAILED
2024-08-26T16:39:45.4077823Z     org.gradle.testkit.runner.UnexpectedBuildFailure: Unexpected build execution failure in C:\Users\RUNNER~1\AppData\Local\Temp\junit-10693879979549064926 with arguments [-PprotobufGradleVersion=0.9.3, -PprotobufVersion=3.23.4, -PkotlinVersion=1.7.20, -PandroidGradleVersion=7.3.0, --configuration-cache, build]
2024-08-26T16:39:45.4080400Z 
2024-08-26T16:39:45.4080586Z     Output:
2024-08-26T16:39:45.4081371Z     Calculating task graph as no cached configuration is available for tasks: build
2024-08-26T16:39:45.4082566Z     > Task :processResources NO-SOURCE
2024-08-26T16:39:45.4083422Z     > Task :processTestResources NO-SOURCE
2024-08-26T16:39:45.4084125Z     > Task :bufFormatCheck
2024-08-26T16:39:45.4084702Z     > Task :bufLint
2024-08-26T16:39:45.4085239Z     > Task :bufGenerate FAILED
2024-08-26T16:39:45.4085603Z 
2024-08-26T16:39:45.4085821Z     FAILURE: Build failed with an exception.
2024-08-26T16:39:45.4086110Z 
2024-08-26T16:39:45.4086221Z     * What went wrong:
2024-08-26T16:39:45.4086652Z     Execution failed for task ':bufGenerate'.
2024-08-26T16:39:45.4089024Z     > > arguments: [D:\a\buf-gradle-plugin\buf-gradle-plugin\build\tmp\test\work\.gradle-test-kit\caches\modules-2\files-2.1\build.buf\buf\1.38.0\d6e207e0d0ca484bbc165b7c10acb85ec0c9eb60\buf-1.38.0-windows-x86_64.exe, generate, --output, C:\Users\runneradmin\AppData\Local\Temp\junit-10693879979549064926\build\bufbuild\generated, --template, C:\Users\runneradmin\AppData\Local\Temp\junit-10693879979549064926\subdir\buf.gen.yaml]
2024-08-26T16:39:45.4091085Z       > exit code: 1
2024-08-26T16:39:45.4091378Z       >    stdout: (empty)
2024-08-26T16:39:45.4091680Z       >    stderr: (below)
2024-08-26T16:39:45.4092530Z       > Failure: Unauthenticated remote plugin rate limit exceeded. For info on how to resolve this, visit: https://buf.build/b/auth-required
2024-08-26T16:39:45.4093296Z       > 
2024-08-26T16:39:45.4093430Z 
2024-08-26T16:39:45.4093436Z 
2024-08-26T16:39:45.4093535Z     * Try:
2024-08-26T16:39:45.4093925Z     > Run with --stacktrace option to get the stack trace.
2024-08-26T16:39:45.4094512Z     > Run with --info or --debug option to get more log output.
2024-08-26T16:39:45.4095032Z     > Run with --scan to get full insights.
2024-08-26T16:39:45.4095462Z     > Get more help at https://help.gradle.org.
```

@pkwarren @drice-buf any ideas on the best way to go about fixing this?

**Comment by pkwarren** [comment]
> @pkwarren @drice-buf any ideas on the best way to go about fixing this?

To avoid rate limits w/ code generation, clients need to be authenticated with the BSR. Alternatively, we could switch to using local generation for the tests instead of relying on remote generation. Looking at how we might do the latter.

**Review by andrewparmet — APPROVED** [review]
Marvelous.
